aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -345,17 +345,13 @@ Tensor _prod(const Tensor &self, int64_t dim_, bool keepdim) {
 
 Tensor& logsumexp_out(Tensor& result, const Tensor &self, int64_t dim_, bool keepdim) {
   int64_t dim = maybe_wrap_dim(dim_, self.dim());
-  // can't take max of empty tensor.
-  if (self.numel() != 0) {
-    auto maxes = at::max_values(self, dim, true);
-    result = at::where((maxes == INFINITY).__or__(maxes == -INFINITY),
-                       maxes,
-                       maxes + at::log(at::sum(at::exp(self - maxes), dim, true)));
-  } else {
-    result = at::log(at::sum(at::exp(self), dim, true));
-  }
-  if (! keepdim)
-    result.squeeze_(dim);
+  // can't take max of empty tensor
+  AT_CHECK(self.numel() != 0, "logsumexp only works on nonempty tensors");
+  auto maxes = at::max_values(self, dim, true);
+  auto maxes_squeezed = (keepdim ? maxes : maxes.squeeze(dim));
+  maxes_squeezed.masked_scatter_(maxes_squeezed.abs() == INFINITY, at::zeros({1}, maxes_squeezed.options()).expand_as(maxes_squeezed));
```

Comment:
Does `at::zeros_like(maxes_squeezed)` work here?

Reply:
What I tried to achieve here is to just have one physical element (and strides = 0).
zeros_like creates a contiguous tensor of the same size.

Reply:
Hmhm. I had previously worked on the result instead of the maxes, here index_fill would be best.

Reply:
Ah that makes sense. `masked_fill_` sounds like a good idea

Reply:
Thanks for insisting that this is unreasonably ugly.